### ISSUE TITLE
Properly integrate report compilation prototype

### ISSUE THREAD

**Issue by sfount** [IMA-WorldHealth/bhima#678]
Over December/ January a prototype to produce clean, well formatted reports was developed, it uses wkhtmltopdf on the server to convert data from the database and a templating method into PDF reports that give us control over page breaking/ size/ orientation/ letterhead etc. 

It was technically proved possible however the implementation in BHIMA has not been properly considered - with proof of concept pages still shown in pages like sale_records and the very brief 2/3 huge button settings page I threw together used for compiling reports. 

As I understand it daily/ weekly/ monthly / custom reports are essential to an accounting software's productivity. Hopefully providing tools that we can really integrate with administration at hospitals as well as reducing work load for audits/ balance evaluations. 

I suggest that the following points are considered to actually implement this feature:
[ ] A list of previously compiled/ generated reports that can be downloaded/ displayed. This will provide a very clear audit trail as well as allowing users to spend a lot of time tweaking the parameters of a report that can then be accessed as many times as required. 
[ ] Standardised creation page - this will follow a similar architecture for most large reports, allowing parameter definition, visual settings, text customisation (titles, description) and validation.
[ ] Design of the relationship between these reports and dynamic data view tools (dashboards/ angular reports etc.) It is important there is a coherent and clear difference in the expected use and functionality of these services
[ ] One design for how this should be implemented in the application were it is purely optional (for example printing a well formed sale invoice)

If this tool is implemented very cleanly I can see it being one of the pillars we demonstrate when trying to sell BHIMA - we know how important the underlying data and collection is but for us to explain that we need to have a clear output.

Please let me know what kind of priority you think this issue is! We can either get started on this straight away or spend time discussing the merits and process we should take.

**Comment by jniles** [comment]
I'll give this a more thorough glance in a minute, but it should be noted the #677  by @DedrickEnc starts the basis of a report configuration service, which is a step towards what you are proposing in the "Standardized Creation Page" section.

**Comment by sfount** [comment]
I know my English can be hard to follow. This is a very quick mockup for how this feature might be implemented in BHIMA. 

![Image of Report Mockup](http://i.imgur.com/pVTGsKC.png)

**Comment by jniles** [comment]
@sfount, did you use mockingbird for that?

**Comment by sfount** [comment]
@jniles https://balsamiq.com/

This tool has saved me hundreds of hours in different projects - I'm sure we can make good use of it here when discussing ideas. 

We could license it for $89 but for everything I've done the free demo instance has been enough so far.

**Comment by jniles** [comment]
Good to know.  Thanks!

**Comment by jniles** [comment]
@sfount, are you proposing that we allow users to create customized reports?  The language under 'Standardized Report Creation' page implies that users are going to be doing far more than simply changing the page orientation and language.  Is that the intent?

Regardless of the answer to the above question, I think this is a very worthy focus for the coming months.  I must confess, I've never actually read the PDF report server side code, so I don't know all the details involved.  However, the UX on the client could certainly use a refresh.  Additionally, having the ability to look at previously generated reports (and generate new ones) would be super useful!

It seems to me, in order to clearly and cleanly create reproducible HTML, PDF, email, etc ... reports, we should first separate out the data logic from the presentation logic.  In the long run, it will be easier to maintain: e.g. you can ask yourself, is this a data issue (bad query, unexpected data), or a presentation issue (unanticipated overflows, too little data breaks rendering)?  That work could be started immediately, with a eye towards eventually allowing the same data query (like the balance sheet or list of employees) to be routed to an HTML renderer or a PDF renderer (or, for that matter, a csv/xlsx render).

Finally, I think this is perhaps a priority for next month.  This will take an in depth discussion, a bit of design work by one or two individuals, and then some heavy duty coding.  The team ought to be together for that, along with Patrick who can help design the initial inventory of reports.

My two cents.

**Comment by sfount** [comment]
@jniles Thanks for the detailed response. 

My intention was not that the user can create custom reports in the sense of a 'report builder'. The idea here is essentially synonymous with addressing a letter - if I'm generating a report for internal use it may be titled differently than a yearly review. 

Looking back one of the main things I aim to take forward is the idea of building the least amount of feature that is really required and putting it together really well. I think having 4 or 5 discrete variables that provide some flexibility on how to address the report would have a positive impact whilst remaining very technically simple. Examples of this would be:
* Title : _optional_
* Subtitle : _optional_ 
* Address To : _optional_ (IMA Accounting? Subsidiary? Chief accountant? Patient?)
* Include letterhead : _optional - default false_ (A template that has already been created by the development team, no need to allow them to specify and create this for now)
* Page orientation 
* Language 
* Compact/ Presentation (if that's appropriate) 

If anything is optional and not defined it simply won't take up any space in the template, essentially leaving just the report. If it's appropriate to allow turning individual columns on a table on/off that is possible but certainly for now I would avoid it. 

I totally agree that any process across BHIMA that needs data for a report should use the same process. Right now as I've mentioned most of the development on proposed reports has just been making PDFs work however I had considered this idea. In the [reports controller](https://github.com/IMA-WorldHealth/bhima/blob/development/server/controllers/reports_proposed/reports.js) you can see that the data is compiled (retrieved) through a 'context', this data is then passed to a 'renderer'. In this case the only renderer that has been written is the PDF renderer, however the idea was this could be passed to a HTML renderer (email) or a client renderer (just send the object). This is very flexible right now as it hasn't been developed very thoroughly but for every architecture decision we should ask if we actually need to support that feature. 

You mentioning that has made me think it that this issue can be broken down into a few different parts. Perhaps it could be developed on its own branch. The nice thing is that a lot of the user facing/ database elements can be done separately from the server architecture. 

Apologies for rambling, looping Patrick in on what reports are essential in what format is an excellent idea.

**Comment by jniles** [comment]
The breakdown of options sound good.  I can't think of any more to help flesh it out.

We have a fairly strong team now.  I see this feature as **the** principle component of bhima from an administrator's standpoint; just like the posting journal is the principle component from the accountant's standpoint.  If we can get (PDF) reports working nicely and seamlessly, the management feedback loop will be in place to drive more work on the accounting side.

I'll move my estimation up.  I personally would be interested in seeing this in bhima by the end of this month.  But I'll defer to @dedrickenc for scheduling this proposal.
